kratos/containers/generic_variables_list.h

```diff
@@ -0,0 +1,305 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Riccardo Rossi
+//                   Jordi Cotela
+//
+
+
+#if !defined(KRATOS_GENERIC_VARIABLES_LIST_H_INCLUDED )
+#define  KRATOS_GENERIC_VARIABLES_LIST_H_INCLUDED
+
+
+// System includes
+#include <string>
+#include <iostream>
+#include "boost/variant.hpp"
+#include "boost/visit_each.hpp"
+
+// External includes
+
+
+// Project includes
+#include "includes/define.h"
+#include "containers/variable.h"
+#include "containers/variable_component.h"
+#include "containers/vector_component_adaptor.h"
+
+namespace Kratos
+{
+
+class VariableNamePrinter : public boost::static_visitor<>
+{
+    public:
+        VariableNamePrinter(std::stringstream& rString) : mrString(rString)
+        {
+        }
+
+        template <class T>
+        void operator()(const T& rVar)
+        {
+            mrString << rVar.Name() << std::endl;
+        }
+
+    private:
+        std::stringstream& mrString;
+};
+
+
+///@addtogroup KratosCore
+///@{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// Generic Container of Variables
+/** This class defines a generic container to hold variables of any type.
+ * it is based on the variant/visitor paradigm.
+ * 
+ * the usage is something like
+ * 
+ *  GenericVariablesList vars;
+ *  vars.AddVariable(TEMPERATURE);
+ *  vars.AddVariable(VELOCITY);
+ *  vars.AddVariable(DISPLACEMENT_X);
+ *  vars.AddVariable(std::string("DISPLACEMENT_Z")); //note that we can add by the name
+ *
+ * operations are defined by creating a custom visitor with an operator specialized per every type
+ * which is then used as
+ *
+ *  MyTestVisitor aux(mp);
+ *   vars.ApplyVisitor( aux );
+*/
+class GenericVariablesList
```

Comment:
maybe we could have variadic templates here allowing to define the type one wants
and define a "using" with the common cases (Variable<double> VariableComponent )